osquery/sql/sqlite_string.cpp

```diff
@@ -152,6 +152,53 @@ static void ip4StringToDecimalFunc(sqlite3_context* context,
   }
 }
 
+/**
+ * @brief Compare two software version strings
+ */
+static void versionLessFunc(sqlite3_context* context,
+                            int argc,
+                            sqlite3_value** argv) {
+  assert(argc == 2);
+
+  if (SQLITE_NULL == sqlite3_value_type(argv[0]) ||
+      SQLITE_NULL == sqlite3_value_type(argv[1])) {
+    sqlite3_result_null(context);
+    return;
+  }
+
+  int major1 = 0, minor1 = 0, patch1 = 0, build1 = 0;
+  std::sscanf((char*)sqlite3_value_text(argv[0]),
```

Comment:
Could you add checks that sscanf actually read that many arguments?

Reply:
I'm not sure how to add checks and still maintain the expected behavior... Take @groob's example: `version_less('0.5.5-22-g85b16ae', '0.5.6')`. These will read different numbers of arguments from `sscanf` but should be comparable with this function.

Reply:
I merged semantic version parser #5343 , may I suggest you to use it here to parse version from the string? Also comparison functionality can be placed as a implementation for operator < or method less to class `SemanticVersion`.

Reply:
A couple notes:
  1. What should be the result if one of the strings are empty?  Currently empty string will equal 0.0.0.0, so version_less("","3.3.1") will equal true.
  2. The build (at least pre-experimental) allows for custom version by specifying OSQUERY_BUILD_VERSION environment variable.  So if I build with OSQUERY_BUILD_VERSION=19.01.1103.0301, then my osquery_info.version="3.3.1-40-gf2a58229_19.01.1103.0301".  This PR as well as the SemanticVersion implementation do not take that into consideration.  I think that's fine, but wanted to mention it.

Reply:
@akindyakov @guliashvili Do you have any plans to update the functionality of the semantic version parser to support the different version strings this supports (based on real-world examples of software versions)? Otherwise, can we decide on a behavior for the empty string and get this merged?